doc/packages/build-support.md

```diff
@@ -0,0 +1,92 @@
+# Build Support {#sec-build-support}
+
+## `pkgs.substitute` {#pkgs-substitute}
+
+`pkgs.substitute` is a wrapper around [the `substitute` function](#fun-substitute) provided by the
+stdenv. It replaces strings in `src` as specified by the `substitutions` argument.
+
+Whereas in a build script, one might call `substitute $infile $outfile --replace-fail @foo@
+${foopkg}/bin/foo`, one might accomplish this with `pkgs.substitute` as such:
+
+```nix
+{ substitute, foopkg }:
+substitute {
+  src = ./sourcefile.txt;
+  substitutions = [
+    "--replace"
+    "@foo@"
+    "${foopkg}/bin/foo"
+  ];
+}
+```
+
+## `pkgs.substituteAll` {#pkgs-substituteall}
+
+`pkgs.substituteAll` is a wrapper around [the `substituteAll` function](#fun-substituteAll) provided
+by the stdenv. It substitutes all instances of `@varName@` in `src` with the value of environment
+variable `varName`. As this uses the `substituteAll` function, its limitations also apply here,
+meaning that variables that you want to replace must:
+
+- start with a *lowercase* letter
+- only contain characters that are alphanumeric or `_`
+
+For example, if `say-goodbye.sh` contained the following:
+
+```
+#! @bash@/bin/bash
+
+echo @unchanged@
+@hello@/bin/hello --greeting @greeting@
+```
+
+the following derivation would make substitutions to `@bash@`, `@hello@`, and `@greeting@`:
+
+```nix
+{
+  substituteAll,
+  bash,
+  hello,
+}:
+substituteAll {
+  src = ./say-goodbye.sh;
+  inherit bash hello;
```

Comment:
It's kind of ill advised to specify environment variable at the top-level, the user could get unexpected results, when accidentially choosing a name that clashes with one of the mkDerivation arguments.
```suggestion
  env = {
    inherit bash hello;
  };
```

Reply:
I'd love more explanation here if you have time -- shouldn't `greeting = "goodbye";` also go into `env` by the same logic?